echopop/nwfsc_feat/biology.py

```diff
@@ -0,0 +1,304 @@
+        is_grouped = False
+        group_cols = []
+    else:
+        # Already a DataFrame from groupby operation
+        regression_df = regression_coefficients.reset_index()
+        # Assign metadata variables required for the output
+        is_grouped = True
+        group_cols = [name for name in regression_coefficients.index.names if name is not None]
+
+    # Initialize fitted weights dataframe
+    weight_fitted_df = length_distribution.copy()
+
+    # Cross merge with regression coefficients
+    weight_fitted_df = weight_fitted_df.merge(regression_df, how="cross").rename(
+        columns={"interval": "length_bin"}
+    )
+
+    # Predict weight per bin using allometric relationship: weight = 10^intercept * length^slope
+    weight_fitted_df["weight_modeled"] = (
+        10.0 ** weight_fitted_df["intercept"] * weight_fitted_df["bin"] ** weight_fitted_df["slope"]
+    )
+
+    # Get the column names if any grouping is required
+    cols = group_cols + ["length_bin"]
+
+    # Quantize weight counts per length bin
+    binned_weight_distribution = (
+        data.groupby(cols, observed=False)["length"].size().fillna(0).to_frame("count")
+    )
+
+    # Quantize weight means per length bin
+    binned_weight_distribution["weight_mean"] = (
+        data.groupby(cols, observed=False)["weight"].mean().fillna(0)
+    )
+
+    # Merge with the fitted weights
+    binned_weight_distribution["weight_modeled"] = weight_fitted_df.set_index(cols)[
+        "weight_modeled"
+    ]
+
+    # Create distribution mask based on imputation settings
+    if impute_bins:
+        # Use modeled weights when count is below threshold
+        if minimum_count_threshold > 0:
+            distribution_mask = binned_weight_distribution["count"] < minimum_count_threshold
+        else:
+            # When threshold is 0, use modeled weights for empty bins only
+            distribution_mask = binned_weight_distribution["count"] == 0
+    else:
+        # No imputation - always use observed means (mask is all False)
+        distribution_mask = pd.Series(False, index=binned_weight_distribution.index)
+
+    # Apply mask to determine final fitted weights
+    binned_weight_distribution["weight_fitted"] = np.where(
+        distribution_mask,
+        binned_weight_distribution["weight_modeled"],
+        binned_weight_distribution["weight_mean"],
+    )
+
+    # Reset index and prepare output
+    result = binned_weight_distribution.reset_index()
+
+    # Mutate, if needed, or otherwise return the pivoted DataFrame
+    if is_grouped:
+        pivot_result = result.pivot(index="length_bin", columns=group_cols, values="weight_fitted")
+        return pivot_result
+    else:
+        if isinstance(result, pd.Series):
+            return result.to_frame("all")
+        else:
+            return result
+
+
+def set_population_metrics(
+    df_nasc: pd.DataFrame,
+    metrics: List[str] = ["abundance", "biomass", "biomass_density"],
```

Comment:
I suggest to remove these options and just calculate everything. It's not a lot of computation and in most cases people would just calculate all of them. This is especially true since you need abundance to get to biomass.